composer/checkpoint/state_dict.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2024 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Useful functions for generating state dicts and manipulating them."""
+
+import fnmatch
+import logging
+from typing import Any, Dict, Optional, Sequence, Union
+
+import torch
+from packaging import version
+from torch import nn
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.nn.parallel import DistributedDataParallel
+
+from composer.models import ComposerModel
+from composer.utils import dist
+
+log = logging.getLogger(__name__)
+
+
+def get_model_state_dict(
+    model: Union[ComposerModel, nn.Module],
+    sharded: bool,
```

Comment:
what if the model is not fsdp but `sharded` is True?  A design question, what's the usecase that the users want to call this function without constructing the trainer object? Or will this API mainly used inside trainer ? I think currently it's not trivial to create a FSDP module outside of the trainer https://github.com/mosaicml/composer/blob/01eec3ad1732074c52cf5da35ef5b8a24531279a/composer/trainer/dist_strategy.py#L212.

Reply:
> what if the model is not fsdp but `sharded` is True? 
It will still return non-sharded. I will add an error

>A design question, what's the usecase that the users want to call this function without constructing the trainer object? Or will this API mainly used inside trainer ? I think currently it's not trivial to create a FSDP module outside of the trainer
This will mostly be used internally, but a user can call `get_model_state_dict(trainer.state.model)`. It is still possible to create an FSDP module outside trainer and we will have advanced users that will do it. See the unit tests for how simple it is to make an FSDP module. Also, once fsdp2 is popularized it will be even easier. Also, I think for the load API, we may aim to add some helper functions to materialize the model

Reply:
thanks! this makes sense!